### PR TITLE
change model to instance, separate solver and instance attributes

### LINKED ISSUE #6: separate model and solver attributes
only 3 overlap, and they don't really make sense (per discussion with @mlubin). so separate these attributes.

also from discussion with @mlubin: define an interface for asking a solver whether a group of "types" of constraints is supported together. and also a way to ask the solver, given an existing model that has already been solved, what types of constraints can be added to it. EDIT moved to #8

there is confusion on my part on what a model is and what a solver is. I think perhaps the terminology could be revisited to make it clearer.

### PR DESCRIPTION
**Pull request by chriscoey**
fixes #6

### CODE REVIEW

**Comment by mlubin** [comment]
I'm leaning towards ``SolverInstance`` for clarity.

**Comment by joaquimg** [comment]
I like 'Instance', but if you need clarity why not 'ModelInstance'?
It seems reasonable, since the one writes a model to describe a mathematical program and when specific numbers are set then we have a instance of a model.
'SolverInstance' might be confuse with 'Solver'.

**Comment by mlubin** [comment]
The important point to communicate is that the MOI "model" should be thought of as the solver's representation of the instance loaded into its API. It's not just the problem data itself, independent of a solver.

**Comment by chriscoey** [comment]
OK I'll change it to `SolverInstance`

File: src/MathOptInterface.jl [hunk 1]
```diff
@@ -5,43 +5,43 @@ module MathOptInterface
...
-    optimize!(m::AbstractModel)
+    optimize!(m::AbstractSolverInstance)
 
 Start the solution procedure.
 """
 function optimize! end
 
 """
-    freemodel!(m::AbstractModel)
+    freesolver instance!(m::AbstractSolverInstance)
```
Comment (mlubin) [inline src/MathOptInterface.jl]: this method name need to be fixed

**Pull request merged**